File: compy_cli/test_dir/python/glfw_test_2.py
import glfw
from compy_python.nones import NULL
from compy_python.strings import to_c_string
from compy_bridge_lib_glfw.d_types import GLFWwindowPtr


def key_callback(
    _window: GLFWwindowPtr, key: int, _scancode: int, action: int, _mods: int
):
    if action == glfw.PRESS:
        print(f"Key {key} pressed")
    elif action == glfw.RELEASE:
        print(f"Key {key} released")


def mouse_button_callback(_window: GLFWwindowPtr, button: int, action: int, _mods: int):
    if action == glfw.PRESS:
        print(f"Mouse button {button} pressed")
    elif action == glfw.RELEASE:
        print(f"Mouse button {button} released")


def cursor_position_callback(_window: GLFWwindowPtr, xpos: float, ypos: float):
    print(f"Mouse moved to ({xpos}, {ypos})")


def glfw_test_2():
    # Initialize the library
    if not glfw.init():
        raise Exception("Failed to initialize GLFW")
    # Create a windowed mode window and its OpenGL context
    window: GLFWwindowPtr = glfw.create_window(
        640, 480, to_c_string("Hello World"), NULL, NULL
    )
    if not window:
        glfw.terminate()
        raise Exception("Failed to create GLFW window")

    # Make the window's context current
    glfw.make_context_current(window)

    # Set input callbacks
    glfw.set_key_callback(window, key_callback)
    glfw.set_mouse_button_callback(window, mouse_button_callback)
    glfw.set_cursor_pos_callback(window, cursor_position_callback)

    # Loop until the user closes the window
    while not glfw.window_should_close(window):
        # Render here, e.g. using pyOpenGL

        # Swap front and back buffers
        glfw.swap_buffers(window)

        # Poll for and process events
        glfw.poll_events()

    glfw.terminate()


if __name__ == "__main__":
    glfw_test_2()

File: pypp_cli/test_dir/cpp/glfw_test_2.cpp
#include "cstdlib"
#include "exceptions/exception.h"
#include "py_str.h"
#include "pypp_util/main_error_handler.h"
#include "pypp_util/print.h"
#include <GLFW/glfw3.h>

void key_callback(GLFWwindow *_window, int key, int _scancode, int action,
                  int _mods) {
    if (action == GLFW_PRESS) {
        print(PyStr(std::format("Key {} pressed", key)));
    } else if (action == GLFW_RELEASE) {
        print(PyStr(std::format("Key {} released", key)));
    }
}

void mouse_button_callback(GLFWwindow *_window, int button, int action,
                           int _mods) {
    if (action == GLFW_PRESS) {
        print(PyStr(std::format("Mouse button {} pressed", button)));
    } else if (action == GLFW_RELEASE) {
        print(PyStr(std::format("Mouse button {} released", button)));
    }
}

void cursor_position_callback(GLFWwindow *_window, double xpos, double ypos) {
    print(PyStr(std::format("Mouse moved to ({}, {})", xpos, ypos)));
}

void glfw_test_2() {
    if (!glfwInit()) {
        throw PyppException(PyStr("Failed to initialize GLFW"));
    }
    GLFWwindow *window = glfwCreateWindow(
        640, 480, PyStr("Hello World").str().c_str(), NULL, NULL);
    if (!window) {
        glfwTerminate();
        throw PyppException(PyStr("Failed to create GLFW window"));
    }
    glfwMakeContextCurrent(window);
    glfwSetKeyCallback(window, key_callback);
    glfwSetMouseButtonCallback(window, mouse_button_callback);
    glfwSetCursorPosCallback(window, cursor_position_callback);
    while (!glfwWindowShouldClose(window)) {
        glfwSwapBuffers(window);
        glfwPollEvents();
    }
    glfwTerminate();
}

int main() {
    try {
        glfw_test_2();
        return 0;
    } catch (...) {
        handle_fatal_exception();
        return EXIT_FAILURE;
    }
}